File: interaction/memories.py
"""
TimeSensei Temporal Memory System
───────────────────────────────────
Manages persistent frozen temporal clones ("memories") with:
  • Crisp, razor-sharp silhouette cropping (zero blur, zero room wall bleed)
  • Strict Depth Ordering: First picture in front, later pictures behind if collided
  • Seamless Environment Lighting Adaptation via hardware LUTs (feels captured in each world)
  • Localized, natural ground contact shadows (no whole-canvas blurring)
  • Precise spatial anchoring (clones stay locked exactly where the visitor posed)
  • Hard collection limit (FIFO eviction at MAX_MEMORIES)
"""
import time
import math
from typing import List, Dict, Any, Optional, Tuple
import numpy as np
import cv2
from config import settings


class FrozenMemory:
    """A stationary frozen temporal clone locked in physical space."""

    _lut_cache: Dict[Tuple, np.ndarray] = {}

    def __init__(
        self,
        crop_bgr: np.ndarray,
        crop_alpha: np.ndarray,
        origin_x: int,
        origin_y: int,
        capture_time: float,
        memory_id: int,
        frame_h: int,
        frame_w: int,
    ):
        self.memory_id = memory_id
        self.capture_time = capture_time
        self.frame_h = frame_h
        self.frame_w = frame_w

        # Razor-sharp RGB cutout and solid alpha mask
        self.crop_bgr = crop_bgr.copy()
        self.crop_alpha = crop_alpha.copy()

        # Spatial Anchoring: Lock clone exactly where visitor was standing
        self.origin_x = float(origin_x)
        self.origin_y = float(origin_y)
        self.height, self.width = crop_bgr.shape[:2]

        # Stationarity
        self.offset_x = 0.0
        self.offset_y = 0.0
        self.rotation = 0.0
        self.shake_intensity = 0.0
        self.is_shattered = False
        self.shatter_finished = False
        self.is_grabbed = False
        self.is_hover_target = False

    @property
    def center_x(self) -> int:
        return int(self.origin_x + self.width / 2.0)

    @property
    def center_y(self) -> int:
        return int(self.origin_y + self.height / 2.0)

    def apply_strike(self, kind: str, direction_x: float = 0.0, direction_y: float = 0.0, force: float = 1.0) -> bool:
        return False

    def update(self):
        """Clone stays stationary."""
        pass

    @classmethod
    def _get_env_lut(cls, env: Any) -> Optional[np.ndarray]:
        """Computes or retrieves cached 1x256x3 hardware Look-Up Table matching active environment."""
        if env is None or getattr(env, "id", "") == "original_room":
            return None

        ambient_tint = getattr(env, "ambient_tint_bgr", (1.0, 1.0, 1.0))
        brightness = getattr(env, "brightness_multiplier", 1.0)
        contrast = getattr(env, "contrast_multiplier", 1.0)
        env_id = getattr(env, "id", "custom")

        key = (env_id, ambient_tint, brightness, contrast)
        lut = cls._lut_cache.get(key)
        if lut is not None:
            return lut

        tint = np.array(ambient_tint, dtype=np.float32)
        intensity = settings.COLOR_MATCH_INTENSITY
        effective_tint = (1.0 - intensity) + (tint * intensity)

        lut_arr = np.zeros((1, 256, 3), dtype=np.uint8)
        for i in range(256):
            for c in range(3):
                val = (i * effective_tint[c] - 128.0) * contrast + 128.0
                lut_arr[0, i, c] = int(np.clip(val * brightness, 0.0, 255.0))

        cls._lut_cache[key] = lut_arr
        return lut_arr

    def render_shadow(self, canvas: np.ndarray):
        """
        Renders localized ground contact shadow under clone base.
        Restricted to a local bounding box under the feet — zero full-canvas blurs.
        """
        ch, cw = canvas.shape[:2]
        dx = int(self.origin_x)
        dy = int(self.origin_y)
        mw, mh = self.width, self.height

        shadow_cx = dx + mw // 2
        shadow_cy = min(ch - 4, dy + mh - 2)
        shadow_rx = int(mw * 0.36)
        shadow_ry = max(5, int(mh * 0.032))

        # Local ROI bounds
        x1 = max(0, shadow_cx - shadow_rx * 2)
        x2 = min(cw, shadow_cx + shadow_rx * 2)
        y1 = max(0, shadow_cy - shadow_ry * 2)
        y2 = min(ch, shadow_cy + shadow_ry * 2)

        if x2 > x1 and y2 > y1 and shadow_cy > 0:
            roi_h, roi_w = y2 - y1, x2 - x1
            local_shadow = np.zeros((roi_h, roi_w), dtype=np.float32)
            local_cx = shadow_cx - x1
            local_cy = shadow_cy - y1

            # Inner dense contact + soft outer falloff
            cv2.ellipse(local_shadow, (local_cx, local_cy), (shadow_rx, shadow_ry), 0, 0, 360, 0.40, -1)
            cv2.ellipse(local_shadow, (local_cx, local_cy + 1), (int(shadow_rx * 1.3), int(shadow_ry * 1.4)), 0, 0, 360, 0.20, -1)
            local_shadow = cv2.GaussianBlur(local_shadow, (15, 15), 0)

            factor = 1.0 - local_shadow[..., None]
            canvas[y1:y2, x1:x2] = (canvas[y1:y2, x1:x2].astype(np.float32) * factor).astype(np.uint8)

    def render_person(self, canvas: np.ndarray, env: Any = None):
        """
        Composites the crisp person cutout onto canvas, dynamically adapted to active environment.
        """
        ch, cw = canvas.shape[:2]
        dx = int(self.origin_x)
        dy = int(self.origin_y)
        mw, mh = self.width, self.height

        # Dynamic environment lighting adaptation
        lut = self._get_env_lut(env)
        if lut is not None:
            adapted_bgr = cv2.LUT(self.crop_bgr, lut)
        elif isinstance(env, tuple) and env != (1.0, 1.0, 1.0):
            # Backward-compat tuple tint
            tint_arr = np.array([env[0], env[1], env[2]], dtype=np.float32)
            intensity = settings.COLOR_MATCH_INTENSITY
            eff = (1.0 - intensity) + (tint_arr * intensity)
            adapted_bgr = np.clip(self.crop_bgr.astype(np.float32) * eff, 0, 255).astype(np.uint8)
        else:
            adapted_bgr = self.crop_bgr

        # Canvas ROI intersection
        src_x1 = max(0, -dx)
        src_y1 = max(0, -dy)
        dst_x1 = max(0, dx)
        dst_y1 = max(0, dy)
        src_x2 = min(mw, cw - dx)
        src_y2 = min(mh, ch - dy)
        dst_x2 = dst_x1 + (src_x2 - src_x1)
        dst_y2 = dst_y1 + (src_y2 - src_y1)

        if src_x1 < src_x2 and src_y1 < src_y2 and dst_x1 < cw and dst_y1 < ch:
            patch_bgr = adapted_bgr[src_y1:src_y2, src_x1:src_x2].astype(np.float32)
            patch_alpha = self.crop_alpha[src_y1:src_y2, src_x1:src_x2, None]
            roi = canvas[dst_y1:dst_y2, dst_x1:dst_x2].astype(np.float32)
            canvas[dst_y1:dst_y2, dst_x1:dst_x2] = (
                patch_bgr * patch_alpha + roi * (1.0 - patch_alpha)
            ).astype(np.uint8)

    def render_onto(self, canvas: np.ndarray, env: Any = None):
        """Unified rendering for single memory."""
        self.render_shadow(canvas)
        self.render_person(canvas, env)


class MemoryManager:
    """Manages all frozen temporal clones in TimeSensei with depth sorting."""

    def __init__(self):
        self.memories: List[FrozenMemory] = []
        self._next_id = 1
        self.grabbed_memory: Optional[FrozenMemory] = None

    def create_memory(self, frame_bgr: np.ndarray, alpha_mask: np.ndarray) -> Optional[FrozenMemory]:
        """
        Creates a razor-sharp frozen clone from the current frame and mask:
          • Rejects all room background bleed (wall, furniture, boxes).
          • Preserves 100% authentic sensor pixel sharpness inside body (no blur/smearing).
          • Anti-aliases the 1-pixel boundary so it blends seamlessly into any virtual world.
        """
        h, w = frame_bgr.shape[:2]

        binary = (alpha_mask > 0.04).astype(np.uint8)
        if not np.any(binary):
            return None

        # Encompass ALL person pixels (including thin fingers, peace signs, and hair)
        bx, by, bw, bh = cv2.boundingRect(binary)
        if bw < 10 or bh < 10:
            return None

        # Generous safety padding (48px) to guarantee zero clipping of fingers or gestures
        pad = 48
        x1 = max(0, bx - pad)
        y1 = max(0, by - pad)
        x2 = min(w, bx + bw + pad)
        y2 = min(h, by + bh + pad)
        bw = x2 - x1
        bh = y2 - y1

        roi_bgr = frame_bgr[y1:y2, x1:x2].copy()
        raw_roi_alpha = alpha_mask[y1:y2, x1:x2].copy()

        # ── Step 1: Smooth S-curve preserving fine fingers, gestures, and hair ──
        # Any confidence > 0.40 becomes solid 1.0 opacity (zero translucency on clothing/body)
        # Any confidence < 0.10 fades smoothly to background
        clean_alpha = np.zeros_like(raw_roi_alpha, dtype=np.float32)
        span = 0.30
        t = np.clip((raw_roi_alpha - 0.10) / span, 0.0, 1.0)
        clean_alpha = t * t * (3.0 - 2.0 * t)

        # ── Step 2: Multi-scale boundary color extension (eliminates black/dark halos) ──
        # Propagate inner body colors into the boundary transition zone (alpha between 0.01 and 0.70)
        # so that edge pixels carry authentic skin/clothing color instead of dark room shadow.
        inner = (clean_alpha >= 0.70).astype(np.uint8)
        border = (clean_alpha > 0.01) & (clean_alpha < 0.70)
        if np.any(inner) and np.any(border):
            k = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (11, 11))
            dilated_colors = cv2.dilate(roi_bgr * inner[..., None], k, iterations=2)
            dilated_weights = cv2.dilate(inner, k, iterations=2)
            valid = (dilated_weights > 0) & border
            if np.any(valid):
                roi_bgr[valid] = dilated_colors[valid]

        # Only clear pixels that are completely transparent
        roi_bgr[clean_alpha == 0] = 0

        # Only feather bottom edge if the person was actually cut off at the bottom frame edge
        if y2 >= h - 2 and bh > 12:
            for r in range(min(6, bh)):
                factor = r / 6.0
                clean_alpha[bh - 1 - r, :] *= factor

        # Enforce MAX_MEMORIES cap (FIFO eviction)
        if len(self.memories) >= settings.MAX_MEMORIES:
            self.memories.pop(0)

        memory = FrozenMemory(
            crop_bgr=roi_bgr,
            crop_alpha=clean_alpha,
            origin_x=x1,
            origin_y=y1,
            capture_time=time.time(),
            memory_id=self._next_id,
            frame_h=h,
            frame_w=w,
        )
        self._next_id += 1
        self.memories.append(memory)
        return memory

    def update_all(self):
        """Update all active clones."""
        for m in self.memories:
            m.update()

    def render_all(self, canvas: np.ndarray, env: Any = None):
        """
        Renders all temporal memories with strict depth ordering:
          1. Floor contact shadows are rendered first (on the ground).
          2. Clones are rendered in reverse chronological order:
             • Later pictures (P2, P3...) are drawn first (in the background).
             • First picture (P1) is drawn last (in the foreground, at first).
             • If collided: first picture stays at first in front, later pictures behind.
             • If not collided: both pictures appear crisp and unhindered.
        """
        if not self.memories:
            return

        # Pass 1: Floor contact shadows underneath all clones
        for m in reversed(self.memories):
            m.render_shadow(canvas)

        # Pass 2: Person cutouts (reversed: earliest clone rendered last = in front)
        for m in reversed(self.memories):
            m.render_person(canvas, env)

    def find_closest_memory(self, px: int, py: int, max_dist: float = 140.0) -> Optional[FrozenMemory]:
        best = None
        best_dist = max_dist
        for m in self.memories:
            dist = math.hypot(px - m.center_x, py - m.center_y)
            if dist < best_dist:
                best_dist = dist
                best = m
        return best

    def update_hover_targets(self, hand_positions: List[Tuple[int, int]]):
        pass

    def try_grab(self, hand_x: int, hand_y: int) -> bool:
        return False

    def drag(self, hand_x: int, hand_y: int):
        pass

    def release(self):
        self.grabbed_memory = None

    def clear_all(self):
        self.memories.clear()
        self._next_id = 1
        self.grabbed_memory = None

File: src/interaction/memories.rs
// TimeSensei Temporal Memory System
// Matches interaction/memories.py exactly.

use std::time::Instant;
use image::{ImageBuffer, Rgb};
use crate::config;

#[derive(Clone)]
pub struct FrozenMemory {
    pub memory_id: usize,
    pub capture_time: Instant,
    pub frame_w: u32,
    pub frame_h: u32,
    pub origin_x: f32,
    pub origin_y: f32,
    pub width: u32,
    pub height: u32,
    pub crop_rgb: ImageBuffer<Rgb<u8>, Vec<u8>>,
    pub crop_alpha: Vec<f32>,
}

impl FrozenMemory {
    pub fn center_x(&self) -> i32 {
        (self.origin_x + self.width as f32 * 0.5) as i32
    }

    pub fn center_y(&self) -> i32 {
        (self.origin_y + self.height as f32 * 0.5) as i32
    }

    pub fn render_shadow(&self, canvas: &mut ImageBuffer<Rgb<u8>, Vec<u8>>) {
        let (cw, ch) = (canvas.width(), canvas.height());
        let dx = self.origin_x as i32;
        let dy = self.origin_y as i32;
        let mw = self.width as i32;
        let mh = self.height as i32;

        let shadow_cx = dx + mw / 2;
        let shadow_cy = (dy + mh - 2).min(ch as i32 - 4);
        let shadow_rx = (mw as f32 * 0.36) as i32;
        let shadow_ry = ((mh as f32 * 0.032) as i32).max(5);

        let x1 = (shadow_cx - shadow_rx * 2).max(0);
        let x2 = (shadow_cx + shadow_rx * 2).min(cw as i32);
        let y1 = (shadow_cy - shadow_ry * 2).max(0);
        let y2 = (shadow_cy + shadow_ry * 2).min(ch as i32);

        if x2 > x1 && y2 > y1 && shadow_cy > 0 {
            for y in y1..y2 {
                for x in x1..x2 {
                    let edx = (x - shadow_cx) as f32 / shadow_rx as f32;
                    let edy = (y - shadow_cy) as f32 / shadow_ry as f32;
                    let d2 = edx * edx + edy * edy;
                    if d2 <= 1.0 {
                        let falloff = (1.0 - d2).clamp(0.0, 1.0) * 0.40;
                        let factor = 1.0 - falloff;
                        let p = canvas.get_pixel_mut(x as u32, y as u32);
                        p.0[0] = (p.0[0] as f32 * factor) as u8;
                        p.0[1] = (p.0[1] as f32 * factor) as u8;
                        p.0[2] = (p.0[2] as f32 * factor) as u8;
                    }
                }
            }
        }
    }

    pub fn render_person(
        &self,
        canvas: &mut ImageBuffer<Rgb<u8>, Vec<u8>>,
        lut: Option<&[[u8; 3]; 256]>,
    ) {
        let (cw, ch) = (canvas.width(), canvas.height());
        let dx = self.origin_x as i32;
        let dy = self.origin_y as i32;
        let mw = self.width as i32;
        let mh = self.height as i32;

        let src_x1 = (-dx).max(0);
        let src_y1 = (-dy).max(0);
        let dst_x1 = dx.max(0);
        let dst_y1 = dy.max(0);
        let src_x2 = mw.min(cw as i32 - dx);
        let src_y2 = mh.min(ch as i32 - dy);

        if src_x1 < src_x2 && src_y1 < src_y2 && dst_x1 < cw as i32 && dst_y1 < ch as i32 {
            let copy_w = (src_x2 - src_x1) as u32;
            let copy_h = (src_y2 - src_y1) as u32;

            for r in 0..copy_h {
                let sy = src_y1 as u32 + r;
                let cy = dst_y1 as u32 + r;
                for c in 0..copy_w {
                    let sx = src_x1 as u32 + c;
                    let cx = dst_x1 as u32 + c;

                    let alpha = self.crop_alpha[(sy * self.width + sx) as usize];
                    if alpha <= 0.001 {
                        continue;
                    }

                    let src_px = self.crop_rgb.get_pixel(sx, sy);
                    let mut adapted_r = src_px.0[0];
                    let mut adapted_g = src_px.0[1];
                    let mut adapted_b = src_px.0[2];

                    if let Some(lut_table) = lut {
                        adapted_r = lut_table[src_px.0[0] as usize][2];
                        adapted_g = lut_table[src_px.0[1] as usize][1];
                        adapted_b = lut_table[src_px.0[2] as usize][0];
                    }

                    let dst_px = canvas.get_pixel_mut(cx, cy);
                    let inv = 1.0 - alpha;
                    dst_px.0[0] = (adapted_r as f32 * alpha + dst_px.0[0] as f32 * inv) as u8;
                    dst_px.0[1] = (adapted_g as f32 * alpha + dst_px.0[1] as f32 * inv) as u8;
                    dst_px.0[2] = (adapted_b as f32 * alpha + dst_px.0[2] as f32 * inv) as u8;
                }
            }
        }
    }
}

pub struct MemoryManager {
    pub memories: Vec<FrozenMemory>,
    next_id: usize,
}

impl MemoryManager {
    pub fn new() -> Self {
        Self {
            memories: Vec::new(),
            next_id: 1,
        }
    }

    pub fn create_memory(
        &mut self,
        frame: &ImageBuffer<Rgb<u8>, Vec<u8>>,
        alpha_mask: &[f32],
    ) -> Option<FrozenMemory> {
        let (fw, fh) = (frame.width() as usize, frame.height() as usize);

        // Find bounding box where alpha > 0.04
        let mut min_x = fw;
        let mut max_x = 0;
        let mut min_y = fh;
        let mut max_y = 0;

        for y in 0..fh {
            let row = y * fw;
            for x in 0..fw {
                if alpha_mask[row + x] > 0.04 {
                    if x < min_x { min_x = x; }
                    if x > max_x { max_x = x; }
                    if y < min_y { min_y = y; }
                    if y > max_y { max_y = y; }
                }
            }
        }

        if max_x <= min_x || max_y <= min_y {
            return None;
        }

        let bw = max_x - min_x;
        let bh = max_y - min_y;
        if bw < 10 || bh < 10 {
            return None;
        }

        // Generous safety padding (48px) to guarantee zero clipping of fingers or gestures
        let pad = 48usize;
        let x1 = min_x.saturating_sub(pad);
        let y1 = min_y.saturating_sub(pad);
        let x2 = (max_x + pad).min(fw);
        let y2 = (max_y + pad).min(fh);
        let crop_w = (x2 - x1) as u32;
        let crop_h = (y2 - y1) as u32;

        let mut crop_rgb = ImageBuffer::new(crop_w, crop_h);
        let mut clean_alpha = vec![0.0f32; (crop_w * crop_h) as usize];

        // Step 1: Smooth Hermite S-curve preserving fine fingers, gestures, and hair
        let span = 0.30f32;
        for cy in 0..crop_h {
            for cx in 0..crop_w {
                let sx = x1 as u32 + cx;
                let sy = y1 as u32 + cy;
                let raw_a = alpha_mask[sy as usize * fw + sx as usize];

                let t = ((raw_a - 0.10) / span).clamp(0.0, 1.0);
                let a = t * t * (3.0 - 2.0 * t);
                let idx = (cy * crop_w + cx) as usize;
                clean_alpha[idx] = a;

                if a > 0.001 {
                    crop_rgb.put_pixel(cx, cy, *frame.get_pixel(sx, sy));
                } else {
                    crop_rgb.put_pixel(cx, cy, Rgb([0, 0, 0]));
                }
            }
        }

        // Step 2: Multi-scale boundary color extension (eliminates black/dark halos)
        // Propagate inner body colors into the boundary transition zone (alpha between 0.01 and 0.70)
        let mut border_pixels = Vec::new();
        for cy in 0..crop_h {
            for cx in 0..crop_w {
                let a = clean_alpha[(cy * crop_w + cx) as usize];
                if a > 0.01 && a < 0.70 {
                    border_pixels.push((cx, cy));
                }
            }
        }

        for (bx, by) in border_pixels {
            // Find nearest inner pixel within radius 5
            let mut found_color = None;
            'search: for r in 1..=5 {
                for dy in -r..=r {
                    for dx in -r..=r {
                        let nx = bx as i32 + dx;
                        let ny = by as i32 + dy;
                        if nx >= 0 && nx < crop_w as i32 && ny >= 0 && ny < crop_h as i32 {
                            let n_idx = (ny as u32 * crop_w + nx as u32) as usize;
                            if clean_alpha[n_idx] >= 0.70 {
                                found_color = Some(*crop_rgb.get_pixel(nx as u32, ny as u32));
                                break 'search;
                            }
                        }
                    }
                }
            }
            if let Some(col) = found_color {
                crop_rgb.put_pixel(bx, by, col);
            }
        }

        // Bottom edge feathering if person was cut off at the bottom frame edge
        if y2 >= fh - 2 && crop_h > 12 {
            for r in 0..6.min(crop_h) {
                let factor = r as f32 / 6.0;
                let row = crop_h - 1 - r;
                for cx in 0..crop_w {
                    clean_alpha[(row * crop_w + cx) as usize] *= factor;
                }
            }
        }

        // FIFO eviction at MAX_MEMORIES
        if self.memories.len() >= config::MAX_MEMORIES {
            self.memories.remove(0);
        }

        let memory = FrozenMemory {
            memory_id: self.next_id,
            capture_time: Instant::now(),
            frame_w: fw as u32,
            frame_h: fh as u32,
            origin_x: x1 as f32,
            origin_y: y1 as f32,
            width: crop_w,
            height: crop_h,
            crop_rgb,
            crop_alpha: clean_alpha,
        };

        self.next_id += 1;
        self.memories.push(memory.clone());
        Some(memory)
    }

    pub fn render_all(
        &self,
        canvas: &mut ImageBuffer<Rgb<u8>, Vec<u8>>,
        lut: Option<&[[u8; 3]; 256]>,
    ) {
        if self.memories.is_empty() {
            return;
        }

        // Pass 1: Floor contact shadows underneath all clones
        for m in self.memories.iter().rev() {
            m.render_shadow(canvas);
        }

        // Pass 2: Person cutouts in strict depth ordering:
        // Later pictures (P2) drawn first (in background)
        // First picture (P1) drawn last (in foreground, at first)
        for m in self.memories.iter().rev() {
            m.render_person(canvas, lut);
        }
    }

    pub fn clear_all(&mut self) {
        self.memories.clear();
        self.next_id = 1;
    }
}
